File: elixir/test/pfid_test.exs
defmodule PfidTest do
  use ExUnit.Case, async: true

  alias Pfid
  alias Pfid.PfidError

  describe "zero/0" do
    test "returns a zero PFID" do
      assert Pfid.zero() == "00000000000000000000000000000000"
    end
  end

  describe "generate/1" do
    test "generates a valid PFID with partition" do
      partition = 123_456_789
      pfid = Pfid.generate(partition)

      assert Pfid.is_pfid?(pfid)
      assert Pfid.extract_partition!(pfid) == partition
    end

    test "generates unique PFIDs" do
      partition = 1
      pfid1 = Pfid.generate(partition)
      pfid2 = Pfid.generate(partition)

      assert pfid1 != pfid2
      assert Pfid.is_pfid?(pfid1)
      assert Pfid.is_pfid?(pfid2)
    end

    test "raises on invalid partition" do
      assert_raise FunctionClauseError, fn ->
        Pfid.generate(-1)
      end

      assert_raise FunctionClauseError, fn ->
        Pfid.generate(1_073_741_824)
      end
    end
  end

  describe "generate/2" do
    test "generates a PFID with partition and timestamp" do
      partition = 123_456_789
      timestamp = 1_234_567_890_000
      pfid = Pfid.generate(partition, timestamp)

      assert Pfid.is_pfid?(pfid)
      assert Pfid.extract_partition!(pfid) == partition
    end

    test "raises on invalid timestamp" do
      assert_raise FunctionClauseError, fn ->
        Pfid.generate(1, -1)
      end

      assert_raise FunctionClauseError, fn ->
        Pfid.generate(1, 281_474_976_710_656)
      end
    end
  end

  describe "generate_example/0" do
    test "generates an example PFID" do
      pfid = Pfid.generate_example()

      assert Pfid.is_pfid?(pfid)
      assert Pfid.extract_partition!(pfid) == 123_456_789
    end
  end

  describe "generate_related/1" do
    test "generates a PFID with the same partition" do
      original_pfid = Pfid.generate(123_456_789)
      related_pfid = Pfid.generate_related(original_pfid)

      assert Pfid.is_pfid?(related_pfid)
      assert Pfid.extract_partition!(related_pfid) == Pfid.extract_partition!(original_pfid)
    end
  end

  describe "generate_root/0" do
    test "generates a PFID with random partition" do
      pfid = Pfid.generate_root()

      assert Pfid.is_pfid?(pfid)
      partition = Pfid.extract_partition!(pfid)
      assert partition >= 0
      assert partition < 1_073_741_824
    end
  end

  describe "generate_binary/1" do
    test "generates a binary PFID" do
      partition = 123_456_789
      binary = Pfid.generate_binary(partition)

      assert byte_size(binary) == 20
      {:ok, encoded} = Pfid.encode(binary)
      assert Pfid.is_pfid?(encoded)
    end
  end

  describe "generate_binary/2" do
    test "generates a binary PFID with timestamp" do
      partition = 123_456_789
      timestamp = 1_234_567_890_000
      binary = Pfid.generate_binary(partition, timestamp)

      assert byte_size(binary) == 20
      {:ok, encoded} = Pfid.encode(binary)
      assert Pfid.is_pfid?(encoded)
    end
  end

  describe "is_pfid?/1" do
    test "returns true for valid PFID" do
      pfid = Pfid.generate(1)
      assert Pfid.is_pfid?(pfid) == true
    end

    test "returns false for invalid strings" do
      assert Pfid.is_pfid?("invalid") == false
      assert Pfid.is_pfid?("") == false
      assert Pfid.is_pfid?("01an4z07byd9df0k79ka1307sr9x4mv") == false
    end

    test "returns false for non-strings" do
      assert Pfid.is_pfid?(123) == false
      assert Pfid.is_pfid?(nil) == false
      assert Pfid.is_pfid?(%{}) == false
    end

    test "returns false for strings starting with 8-9" do
      assert Pfid.is_pfid?("8" <> String.duplicate("0", 31)) == false
      assert Pfid.is_pfid?("9" <> String.duplicate("0", 31)) == false
    end
  end

  describe "encode/1" do
    test "encodes a valid binary" do
      partition = 123_456_789
      binary = Pfid.generate_binary(partition)
      {:ok, encoded} = Pfid.encode(binary)

      assert Pfid.is_pfid?(encoded)
      assert Pfid.extract_partition!(encoded) == partition
    end

    test "returns error for invalid binary size" do
      {:error, %PfidError{code: :invalid_binary}} = Pfid.encode(<<1, 2, 3>>)
    end

    test "returns error for non-binary" do
      {:error, %PfidError{code: :invalid_binary}} = Pfid.encode("not binary")
      {:error, %PfidError{code: :invalid_binary}} = Pfid.encode(123)
    end
  end

  describe "decode/1" do
    test "decodes a valid PFID" do
      partition = 123_456_789
      pfid = Pfid.generate(partition)
      {:ok, binary} = Pfid.decode(pfid)

      assert byte_size(binary) == 20
      {:ok, encoded} = Pfid.encode(binary)
      assert encoded == pfid
    end

    test "returns error for invalid PFID" do
      {:error, %PfidError{code: :invalid_pfid}} = Pfid.decode("invalid")
      {:error, %PfidError{code: :invalid_pfid}} = Pfid.decode("")
    end

    test "returns error for non-string" do
      {:error, %PfidError{code: :invalid_pfid}} = Pfid.decode(123)
    end
  end

  describe "extract_partition/1" do
    test "extracts partition from valid PFID" do
      partition = 123_456_789
      pfid = Pfid.generate(partition)
      {:ok, extracted} = Pfid.extract_partition(pfid)

      assert extracted == partition
    end

    test "returns error for invalid PFID" do
      {:error, %PfidError{code: :invalid_pfid}} = Pfid.extract_partition("invalid")
    end
  end

  describe "extract_partition!/1" do
    test "extracts partition from valid PFID" do
      partition = 123_456_789
      pfid = Pfid.generate(partition)

      assert Pfid.extract_partition!(pfid) == partition
    end

    test "raises on invalid PFID" do
      assert_raise PfidError, fn ->
        Pfid.extract_partition!("invalid")
      end
    end
  end

  describe "generate_partition/0" do
    test "generates a valid partition" do
      partition = Pfid.generate_partition()

      assert partition >= 0
      assert partition < 1_073_741_824
    end

    test "generates different partitions" do
      partitions = for _ <- 1..10, do: Pfid.generate_partition()
      unique_partitions = Enum.uniq(partitions)

      # Very unlikely to have duplicates, but possible
      assert length(unique_partitions) >= 8
    end
  end

  describe "round-trip encoding/decoding" do
    test "encode and decode are inverse operations" do
      partition = 123_456_789
      pfid = Pfid.generate(partition)

      {:ok, binary} = Pfid.decode(pfid)
      {:ok, encoded} = Pfid.encode(binary)

      assert encoded == pfid
    end

    test "binary encode and decode are inverse operations" do
      partition = 123_456_789
      binary = Pfid.generate_binary(partition)

      {:ok, encoded} = Pfid.encode(binary)
      {:ok, decoded} = Pfid.decode(encoded)

      assert decoded == binary
    end
  end

  describe "partition consistency" do
    test "all generated PFIDs with same partition have same partition" do
      partition = 123_456_789

      pfids =
        for _ <- 1..10 do
          Pfid.generate(partition)
        end

      partitions = Enum.map(pfids, &Pfid.extract_partition!/1)

      assert Enum.all?(partitions, fn p -> p == partition end)
    end
  end
end

File: typescript/test/pfid.test.ts
import { describe, test, expect } from 'vitest';
import {
  zero,
  generate,
  generateWithTimestamp,
  generateExample,
  generateRelated,
  generateRoot,
  generateBinary,
  generateBinaryWithTimestamp,
  isPfid,
  encode,
  decode,
  extractPartition,
  generatePartition,
  PfidError,
} from '../src';

describe('zero', () => {
  test('returns a zero PFID', () => {
    expect(zero()).toBe('00000000000000000000000000000000');
  });
});

describe('generate', () => {
  test('generates a valid PFID with partition', () => {
    const partition = 123_456_789;
    const pfid = generate(partition);

    expect(isPfid(pfid)).toBe(true);
    expect(extractPartition(pfid)).toBe(partition);
  });

  test('generates unique PFIDs', () => {
    const partition = 1;
    const pfid1 = generate(partition);
    const pfid2 = generate(partition);

    expect(pfid1).not.toBe(pfid2);
    expect(isPfid(pfid1)).toBe(true);
    expect(isPfid(pfid2)).toBe(true);
  });

  test('throws on invalid partition', () => {
    expect(() => generate(-1)).toThrow();
    expect(() => generate(1_073_741_824)).toThrow();
  });
});

describe('generateWithTimestamp', () => {
  test('generates a PFID with partition and timestamp', () => {
    const partition = 123_456_789;
    const timestamp = 1_234_567_890_000;
    const pfid = generateWithTimestamp(partition, timestamp);

    expect(isPfid(pfid)).toBe(true);
    expect(extractPartition(pfid)).toBe(partition);
  });

  test('throws on invalid timestamp', () => {
    expect(() => generateWithTimestamp(1, -1)).toThrow();
    expect(() => generateWithTimestamp(1, 281_474_976_710_656)).toThrow();
  });
});

describe('generateExample', () => {
  test('generates an example PFID', () => {
    const pfid = generateExample();

    expect(isPfid(pfid)).toBe(true);
    expect(extractPartition(pfid)).toBe(123_456_789);
  });
});

describe('generateRelated', () => {
  test('generates a PFID with the same partition', () => {
    const originalPfid = generate(123_456_789);
    const relatedPfid = generateRelated(originalPfid);

    expect(isPfid(relatedPfid)).toBe(true);
    expect(extractPartition(relatedPfid)).toBe(
      extractPartition(originalPfid)
    );
  });
});

describe('generateRoot', () => {
  test('generates a PFID with random partition', () => {
    const pfid = generateRoot();

    expect(isPfid(pfid)).toBe(true);
    const partition = extractPartition(pfid);
    expect(partition).toBeGreaterThanOrEqual(0);
    expect(partition).toBeLessThan(1_073_741_824);
  });
});

describe('generateBinary', () => {
  test('generates a binary PFID', () => {
    const partition = 123_456_789;
    const binary = generateBinary(partition);

    expect(binary.length).toBe(20);
    const encoded = encode(binary);
    expect(isPfid(encoded)).toBe(true);
  });
});

describe('generateBinaryWithTimestamp', () => {
  test('generates a binary PFID with timestamp', () => {
    const partition = 123_456_789;
    const timestamp = 1_234_567_890_000;
    const binary = generateBinaryWithTimestamp(partition, timestamp);

    expect(binary.length).toBe(20);
    const encoded = encode(binary);
    expect(isPfid(encoded)).toBe(true);
  });
});

describe('isPfid', () => {
  test('returns true for valid PFID', () => {
    const pfid = generate(1);
    expect(isPfid(pfid)).toBe(true);
  });

  test('returns false for invalid strings', () => {
    expect(isPfid('invalid')).toBe(false);
    expect(isPfid('')).toBe(false);
    expect(isPfid('01an4z07byd9df0k79ka1307sr9x4mv')).toBe(false);
  });

  test('returns false for non-strings', () => {
    expect(isPfid(123 as any)).toBe(false);
    expect(isPfid(null as any)).toBe(false);
    expect(isPfid({} as any)).toBe(false);
  });

  test('returns false for strings starting with 8-9', () => {
    expect(isPfid('8' + '0'.repeat(31))).toBe(false);
    expect(isPfid('9' + '0'.repeat(31))).toBe(false);
  });
});

describe('encode', () => {
  test('encodes a valid binary', () => {
    const partition = 123_456_789;
    const binary = generateBinary(partition);
    const encoded = encode(binary);

    expect(isPfid(encoded)).toBe(true);
    expect(extractPartition(encoded)).toBe(partition);
  });

  test('throws error for invalid binary size', () => {
    expect(() => encode(Buffer.from([1, 2, 3]))).toThrow(PfidError);
    expect(() => encode(Buffer.from([1, 2, 3]))).toThrow(/invalid binary/);
  });

  test('throws error for non-binary', () => {
    expect(() => encode('not binary' as any)).toThrow(PfidError);
    expect(() => encode(123 as any)).toThrow(PfidError);
  });
});

describe('decode', () => {
  test('decodes a valid PFID', () => {
    const partition = 123_456_789;
    const pfid = generate(partition);
    const binary = decode(pfid);

    expect(binary.length).toBe(20);
    const encoded = encode(binary);
    expect(encoded).toBe(pfid);
  });

  test('throws error for invalid PFID', () => {
    expect(() => decode('invalid')).toThrow(PfidError);
    expect(() => decode('invalid')).toThrow(/invalid PFID/);
    expect(() => decode('')).toThrow(PfidError);
  });

  test('throws error for non-string', () => {
    expect(() => decode(123 as any)).toThrow(PfidError);
  });
});

describe('extractPartition', () => {
  test('extracts partition from valid PFID', () => {
    const partition = 123_456_789;
    const pfid = generate(partition);
    const extracted = extractPartition(pfid);

    expect(extracted).toBe(partition);
  });

  test('throws error for invalid PFID', () => {
    expect(() => extractPartition('invalid')).toThrow(PfidError);
    expect(() => extractPartition('invalid')).toThrow(/invalid PFID/);
  });
});

describe('generatePartition', () => {
  test('generates a valid partition', () => {
    const partition = generatePartition();

    expect(partition).toBeGreaterThanOrEqual(0);
    expect(partition).toBeLessThan(1_073_741_824);
  });

  test('generates different partitions', () => {
    const partitions: number[] = [];
    for (let i = 0; i < 10; i++) {
      partitions.push(generatePartition());
    }
    const uniquePartitions = [...new Set(partitions)];

    // Very unlikely to have duplicates, but possible
    expect(uniquePartitions.length).toBeGreaterThanOrEqual(8);
  });
});

describe('round-trip encoding/decoding', () => {
  test('encode and decode are inverse operations', () => {
    const partition = 123_456_789;
    const pfid = generate(partition);

    const binary = decode(pfid);
    const encoded = encode(binary);
    expect(encoded).toBe(pfid);
  });

  test('binary encode and decode are inverse operations', () => {
    const partition = 123_456_789;
    const binary = generateBinary(partition);

    const encoded = encode(binary);
    const decoded = decode(encoded);
    expect(decoded).toEqual(binary);
  });
});

describe('partition consistency', () => {
  test('all generated PFIDs with same partition have same partition', () => {
    const partition = 123_456_789;

    const pfids: string[] = [];
    for (let i = 0; i < 10; i++) {
      pfids.push(generate(partition));
    }

    const partitions = pfids.map((pfid) => extractPartition(pfid));

    expect(partitions.every((p) => p === partition)).toBe(true);
  });
});
